### ISSUE TITLE
com.hazelcast.jet.core.GracefulShutdownTest.when_shutdownGracefulWhileRestartGraceful_then_restartsFromTerminalSnapshot

### ISSUE THREAD

**Issue by cangencer** [hazelcast/hazelcast-jet#2166]
http://jenkins.hazelcast.com/job/jet-oss-pr-builder/2637/testReport/junit/com.hazelcast.jet.core/GracefulShutdownTest/when_shutdownGracefulWhileRestartGraceful_then_restartsFromTerminalSnapshot/

Error Message

test timed out after 300000 milliseconds

Stacktrace

```
org.junit.runners.model.TestTimedOutException: test timed out after 300000 milliseconds
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947)
	at com.hazelcast.jet.Job.join(Job.java:137)
	at com.hazelcast.jet.core.GracefulShutdownTest.when_shutdownGracefulWhileRestartGraceful_then_restartsFromTerminalSnapshot(GracefulShutdownTest.java:206)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by viliam-durina** [comment]
I investigated the problem, here are the findings: The test starts with 2 members. Then initiates a graceful job restart, but the snapshot map is configured with a MapStore that blocks. So the execution does not terminate because the terminal snapshot is blocked. The test asserts that the blocking happened and then continues to gracefully shutdown the 2nd member (non-master) in a spawned thread to not block the test until the shutdown is done. Then it unblocks the MapStore and asserts that the shutdown finished. The graceful shutdown is supposed to start a terminal snapshot, but since one is already started, it should not. Then the test joins the job and this is where it got stuck.

From the logs we see the terminal snapshot, then the shutdown and then we see that the member terminated successfully. We see that the 2nd member was added to `shuttingDownMembers` on the master and that the job does not restart because of that:
```
2020-04-14 14:58:26,917 [DEBUG] [hz.thirsty_montalcini.cached.thread-8] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.1-SNAPSHOT] Rescheduling restart of '0432-8b8f-1c80-0001': cluster is not safe
```
That scheduled restart should happen after 2 seconds. We see the member shut down and being removed from `shuttingDownMembers`. The last log message is from ~1.5 secs after the scheduled restart, after that there are only messages from other regular activity such as "Job cleanup" and "health monitor" (triggered by heap usage >70%).

I suspect that the scheduled action wasn't executed. All paths there lead to some log message, but none is found. But I might have missed a path, it's way too easy to suspect a java service to just lose a scheduled action.

Also the test failed for the first time after maybe 2 years, so I didn't try to reproduce it locally.
